### PR TITLE
wso2ei 6.4.0 (new formula)

### PR DESCRIPTION
**Pull request by irunika**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
github release: https://github.com/wso2/product-ei/releases/tag/v6.4.0

### CODE REVIEW

**Comment by fxcoudert** [comment]
Version 6.4.0 does not appear to be released yet on GitHub: https://github.com/wso2/product-ei/releases
We would prefer using github links

File: Formula/wso2ei-6.4.0.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  bottle :unneeded
+  depends_on :java => "1.8"
+
+  def install
+    product_name = "wso2ei"
+    product_version = "6.4.0"
+    profiles = %w[integrator business-process broker analytics-dashboard analytics-worker micro-integrator msf4j]
+
+    puts "\nInstalling WSO2 Enterprise Integrator #{product_version}..."
```
Comment (fxcoudert) [inline Formula/wso2ei-6.4.0.rb]: No `puts`. The whole process should be silent.

File: Formula/wso2ei-6.4.0.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    puts "\nRun the following commands to start each profile of the WSO2 Enterprise Integrator."
+    profiles.each do |profile|
+      puts "\t#{profile}".ljust(20) + ": #{product_name}-#{product_version}-#{profile}"
+    end
+    puts "\nCheers!!"
+    puts
+  end
+
+  test do
+    system "true"
```
Comment (fxcoudert) [inline Formula/wso2ei-6.4.0.rb]: We need a test that exercises the some of the functionality of the app.

File: Formula/wso2ei-6.4.0.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    profiles.each do |profile|
+      chmod 0755, "bin/#{product_name}-#{product_version}-#{profile}"
+      bin.install "bin/#{product_name}-#{product_version}-#{profile}"
+    end
+
+    libexec.install Dir["*"]
+    bin.env_script_all_files(libexec/"bin", Language::Java.java_home_env("1.8"))
+
+    puts "Installation completed."
+    puts "\nRun the following commands to start each profile of the WSO2 Enterprise Integrator."
```
Comment (fxcoudert) [inline Formula/wso2ei-6.4.0.rb]: If it is specific to Homebrew, this information should go in the `caveats` section

File: Formula/wso2ei-6.4.0.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  depends_on :java => "1.8"
+
+  def install
+    product_name = "wso2ei"
+    product_version = "6.4.0"
+    profiles = %w[integrator business-process broker analytics-dashboard analytics-worker micro-integrator msf4j]
+
+    puts "\nInstalling WSO2 Enterprise Integrator #{product_version}..."
+    profiles.each do |profile|
+      chmod 0755, "bin/#{product_name}-#{product_version}-#{profile}"
```
Comment (fxcoudert) [inline Formula/wso2ei-6.4.0.rb]: Why do these files have the wrong permission?

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/wso2ei-6.4.0.rb (hunk 1 above)
Comment (irunika) [reply Formula/wso2ei-6.4.0.rb]: The reason why we chose brew is to give users a seamless experience.
So we need to print a message after installation is completed to try the product for the first time.
Where can we put it?

File: Formula/wso2ei-6.4.0.rb (hunk 2 above)
Comment (irunika) [reply Formula/wso2ei-6.4.0.rb]: With this brew package, we want to just install the WSO2 runtime in the user's filesystem. To make the runtime functional the user has to "start" the server. We do not start the server as part of the installation and give the control of that to the user.

To test the installation we can check for an internal config file within the runtime. Hope that's sufficient to comply with Homebrew standards.

Thanks!

File: Formula/wso2ei-6.4.0.rb (hunk 4 above)
Comment (irunika) [reply Formula/wso2ei-6.4.0.rb]: Could you please elaborate more on this permission issue. This is sufficient permission to start the WSO2 runtime.

Comment (fxcoudert) [reply Formula/wso2ei-6.4.0.rb]: Why the need for `chmod`? The files should have the required permissions in the archive, no?

File: Formula/wso2ei-6.4.0.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+class Wso2ei640 < Formula
```
Comment (commitay) [inline Formula/wso2ei-6.4.0.rb]: This needs to be renamed so it doesn't include the version.

This needs to be done with all of the open PR.

**Review by commitay — CHANGES_REQUESTED** [review]

File: Formula/wso2ei-6.4.0.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    PROFILES.each do |profile|
+      bin.install "bin/#{PRODUCT_NAME}-#{PRODUCT_VERSION}-#{profile}"
+    end
+
+    libexec.install Dir["*"]
+    bin.env_script_all_files(libexec/"bin", Language::Java.java_home_env("1.8"))
+  end
+
+  def caveats
+    caveat_msg = "Run the following commands to start each profile of the WSO2 Enterprise Integrator.\n"
```
Comment (commitay) [inline Formula/wso2ei-6.4.0.rb]: This isn't Homebrew specific information, it should be removed.

File: Formula/wso2ei-6.4.0.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+class Wso2ei640 < Formula
+  PRODUCT_NAME = "wso2ei".freeze
```
Comment (commitay) [inline Formula/wso2ei-6.4.0.rb]: Inline this.

File: Formula/wso2ei-6.4.0.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  PRODUCT_VERSION = "6.4.0".freeze
```
Comment (commitay) [inline Formula/wso2ei-6.4.0.rb]: Inline this.

File: Formula/wso2ei-6.4.0.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  PROFILES = %w[integrator business-process broker analytics-dashboard analytics-worker micro-integrator msf4j].freeze
```
Comment (commitay) [inline Formula/wso2ei-6.4.0.rb]: Move this into the install.

File: Formula/wso2ei-6.4.0.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  homepage "https://wso2.com/integration/"
+  url "https://bintray.com/wso2/binary/download_file?file_path=#{PRODUCT_NAME}-#{PRODUCT_VERSION}.zip"
+  sha256 "36ab3982cc517674ab1673c1947093bf8022e1f4fb971a2fe6fa1b3085a13917"
+
+  bottle :unneeded
+  depends_on :java => "1.8"
+
+  def install
+    PROFILES.each do |profile|
+      bin.install "bin/#{PRODUCT_NAME}-#{PRODUCT_VERSION}-#{profile}"
```
Comment (commitay) [inline Formula/wso2ei-6.4.0.rb]: Use `#{version}` instead of `#{PRODUCT_VERSION}`

**Comment by commitay** [comment]
@irunika All of the comments on this PR need to be applied to the other open PRs.

File: Formula/wso2ei-6.4.0.rb (hunk 6 above)
Comment (irunika) [reply Formula/wso2ei-6.4.0.rb]: Can you elaborate more on this issue? Do you mean that this information is wrong to put in the caveats?

Comment (commitay) [reply Formula/wso2ei-6.4.0.rb]: Yes.

Comment (irunika) [reply Formula/wso2ei-6.4.0.rb]: So if we need to give post-installation information, is there a place put it in Homebrew?

File: Formula/wso2ei-6.4.0.rb (hunk 7 above)
Comment (irunika) [reply Formula/wso2ei-6.4.0.rb]: What do you mean by inlining here?

File: Formula/wso2ei-6.4.0.rb (hunk 6 above)
Comment (commitay) [reply Formula/wso2ei-6.4.0.rb]: That is the perview of upstream documentation.

File: Formula/wso2ei-6.4.0.rb (hunk 7 above)
Comment (commitay) [reply Formula/wso2ei-6.4.0.rb]: `"bin/wsoei-#{version}-#{profile}"`

File: Formula/wso2ei-6.4.0.rb (hunk 5 above)
Comment (irunika) [reply Formula/wso2ei-6.4.0.rb]: Isn't this preventing us from using the package name as `wso2ei-6.4.0`?
We have a requirement to install the  products as `brew install wso2ei-#{version}`

Comment (commitay) [reply Formula/wso2ei-6.4.0.rb]: The formula should able to be installed as `brew install wso2ei`.

If having them versioned is important I'd suggest [creating your own tap.](https://github.com/Homebrew/brew/blob/master/docs/How-to-Create-and-Maintain-a-Tap.md)

**Comment by commitay** [comment]
```ruby
class Wso2ei < Formula
  desc "WSO2 Enterprise Integrator"
  homepage "https://wso2.com/integration/"
  url "https://dl.bintray.com/wso2/binary/wso2ei-6.4.0.zip"
  sha256 "36ab3982cc517674ab1673c1947093bf8022e1f4fb971a2fe6fa1b3085a13917"

  bottle :unneeded

  depends_on :java => "1.8"

  def install
    %w[
      integrator
      business-process
      broker
      analytics-dashboard
      analytics-worker
      micro-integrator
      msf4j
    ].each do |profile|
      bin.install "bin/wso2ei-#{version}-#{profile}" => "wso2ei-#{profile}"
    end

    libexec.install Dir["*"]
    bin.env_script_all_files(libexec/"bin", Language::Java.java_home_env("1.8"))
  end

  test do
    system "true"
  end
end
```

**Comment by irunika** [comment]
@commitay & @fxcoudert 
Can we please check this now?

File: Formula/wso2ei.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
+class Wso2ei < Formula
+  desc "WSO2 Enterprise Integrator 6.4.0"
+  homepage "https://wso2.com/integration/"
+  url "https://bintray.com/wso2/binary/download_file?file_path=wso2ei-6.4.0.zip"
+  sha256 "36ab3982cc517674ab1673c1947093bf8022e1f4fb971a2fe6fa1b3085a13917"
+
+  bottle :unneeded
+  depends_on :java => "1.8"
```
Comment (commitay) [inline Formula/wso2ei.rb]: Newline between these two.

File: Formula/wso2ei.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
+class Wso2ei < Formula
+  desc "WSO2 Enterprise Integrator 6.4.0"
+  homepage "https://wso2.com/integration/"
+  url "https://bintray.com/wso2/binary/download_file?file_path=wso2ei-6.4.0.zip"
```
Comment (commitay) [inline Formula/wso2ei.rb]: `https://dl.bintray.com/wso2/binary/wso2ei-6.4.0.zip` or a GitHub link.

File: Formula/wso2ei.rb [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  depends_on :java => "1.8"
+
+  def install
+    %w[
+      integrator
+      business-process
+      broker
+      analytics-dashboard
+      analytics-worker
+      micro-integrator msf4j
```
Comment (commitay) [inline Formula/wso2ei.rb]: Only one per line.

File: Formula/wso2ei.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
+class Wso2ei < Formula
+  desc "WSO2 Enterprise Integrator 6.4.0"
```
Comment (commitay) [inline Formula/wso2ei.rb]: Remove the version from the description.

File: Formula/wso2ei.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  def install
+    %w[
+      integrator
+      business-process
+      broker
+      analytics-dashboard
+      analytics-worker
+      micro-integrator msf4j
+    ].each do |profile|
+      bin.install "bin/wso2ei-6.4.0-#{profile}" => "wso2ei-#{profile}"
```
Comment (commitay) [inline Formula/wso2ei.rb]: `"bin/wso2ei-#{version}-#{profile}

File: Formula/wso2ei.rb (hunk 12 above)
Comment (irunika) [reply Formula/wso2ei.rb]: We want to keep the Bintray URL.

File: Formula/wso2ei.rb [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  homepage "https://wso2.com/integration/"
+  url "https://bintray.com/wso2/binary/download_file?file_path=wso2ei-6.4.0.zip"
+  sha256 "36ab3982cc517674ab1673c1947093bf8022e1f4fb971a2fe6fa1b3085a13917"
+
+  bottle :unneeded
+
+  depends_on :java => "1.8"
+
+  def install
+    version = "6.4.0"
```
Comment (commitay) [inline Formula/wso2ei.rb]: This isn't needed, it's detected automatically.

File: Formula/wso2ei.rb (hunk 11 above)
Comment (irunika) [reply Formula/wso2ei.rb]: Done

File: Formula/wso2ei.rb (hunk 12 above)
Comment (commitay) [reply Formula/wso2ei.rb]: Please change this to `https://dl.bintray.com/wso2/binary/wso2ei-6.4.0.zip`, it follows the standard pattern that we use for Bintray URLs.

File: Formula/wso2ei.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
+class Wso2ei < Formula
+  desc "WSO2 Enterprise Integrator"
+  homepage "https://wso2.com/integration/"
+  url "https://bintray.com/wso2/binary/download_file?file_path=wso2ei-6.4.0.zip"
```
Comment (commitay) [inline Formula/wso2ei.rb]: Please change this to `https://dl.bintray.com/wso2/binary/wso2ei-6.4.0.zip`, it follows the standard pattern that we use for Bintray URLs.

**Review by commitay — CHANGES_REQUESTED** [review]

File: Formula/wso2ei.rb [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    ].each do |profile|
+      bin.install "bin/wso2ei-#{version}-#{profile}" => "wso2ei-#{profile}"
+    end
+
+    libexec.install Dir["*"]
+    bin.env_script_all_files(libexec/"bin", Language::Java.java_home_env("1.8"))
+  end
+
+  test do
+    system "true"
```
Comment (commitay) [inline Formula/wso2ei.rb]: Just needs a test now.

Comment (irunika) [reply Formula/wso2ei.rb]: With this brew package, we want to just install the WSO2 runtime in the user's filesystem. To make the runtime functional the user has to "start" the server. We do not start the server as part of the installation and give the control of that to the user.

To test the installation we can check for an internal config file within the runtime. Hope that's sufficient to comply with Homebrew standards.

Thanks!

Comment (commitay) [reply Formula/wso2ei.rb]: The test needs to exercise and verify the functionality of the software. It runs sandboxed on CI, not on users systems.

**Comment by irunika** [comment]
Even though I run this locally it fails. However, if I run the alias, the server starts well. Any idea why it only fails when trying to start through brew test?

**Comment by irunika** [comment]
@commitay & @fxcoudert 
Done all the changes and added test case needed. Please check.

Thanks!

File: Formula/wso2ei.rb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+      bin.install "bin/wso2ei-#{version}-#{profile}" => "wso2ei-#{profile}"
+    end
+
+    libexec.install Dir["*"]
+    bin.env_script_all_files(libexec/"bin", Language::Java.java_home_env("1.8"))
+  end
+
+  test do
+    ENV["CARBON_HOME"] = testpath
+    cp_r Dir["#{libexec}/*"], testpath
```
Comment (fxcoudert) [inline Formula/wso2ei.rb]: We shouldn't be copying the entire installation into the testpath. It should run from where it is installed.

Comment (irunika) [reply Formula/wso2ei.rb]: We have a challenge when running tests against the install directory as it depends on OpenJDK we ship with our package. We followed the same practice followed by Apache Tomcat brew package (https://github.com/Homebrew/homebrew-core/blob/master/Formula/tomcat.rb).

If Tomcat way is wrong, could you please suggest a better way to run the test against a different JDK?

Thanks!

Comment (irunika) [reply Formula/wso2ei.rb]: We are making the announcement of choosing Homebrew as the package distributor for WSO2 runtimes. We really want to host these packages ASAP before the pubic announce is made. I really apreciate your support on this.

Thanks!

**Comment by claui** [comment]
@fxcoudert The [6.4.0 release](https://github.com/wso2/product-ei/releases/tag/v6.4.0) seems to be on GitHub now.

**Comment by irunika** [comment]
Yes, you can find it in https://github.com/wso2/product-ei/releases/tag/v6.4.0.

Can we please merge PRs from WSO2 if everything is resolved since we have already done the requested changes to the other PRs too?

Also please consider that we are already behind the schedule to do our public announcement due to the delay of getting the PRs for Homebrew merged. We realy appreciate your support on this.

Thanks!

**Comment by commitay** [comment]
> OpenJDK we ship with our package

This is against Homebrew policy for `core` formulae, it shouldn't bundle a JDK.

I'd suggest [creating your own tap](https://github.com/Homebrew/brew/blob/master/docs/How-to-Create-and-Maintain-a-Tap.md) for these formulae.

**Comment by commitay** [comment]
> Also please consider that we are already behind the schedule to do our public announcement due to the delay of getting the PRs for Homebrew merged.

We're volunteers working on this in our free time.

**Pull request closed without merging**